[토박][bakha, TOKO] mailbox - 첫번째 PR

aaaz425 opened the pull request:
# 미리보기
<img src="https://user-images.githubusercontent.com/101464713/229046866-2b83c9a0-f85f-4383-b26e-0011db8955cc.mov">

## 페어 프로그래밍 정리

- 1일차 (월)
  - 간단히 페어 프로그래밍의 방향을 잡았다
  - git 으로 프로젝트를 공유할 방법에 대해 토론했다
  - html, css를 간단히 작업하고 JS 를 아주 조금 다뤘다
- 2일차 (화)
  - JS 기능에 대한 방향을 잡았다
  - 본격적인 구현을 시도했다 -> 실패
    - 구현 실패 원인 분석
    ```
    1. 요구사항을 작게 나누었지만, 막상 구현할 때는 여러가지 기능을 한번에 다 넣으려고 했다
    2. driver와 navigator의 역할 분담 실패 - 각자 driver의 역할을 맡을 때 너무 수동적으로 임함 -> navigator의 부담이 증가했다
    ```
- 3일차 (수)
  - 미션 해결을 위해 개인 공부시간을 가졌다
  - 구현 방식에 대한 방향에 대해 토론했다
  - 각자 생각한 구현 방식이 달랐다 -> 의견 조율 실패
    - 의견 조율 실패 이유
    ```
    1. 의견을 나눌 시간이 부족했다
    2. 서로 미션 구현을 위한 구체적인 계획이 부족했다
    3. toko : 현재 상황에 맞추어 미션의 내용을 축소 및 변경하자
    4. bakha : 미션의 핵심(노드 탐색, 재귀를 통한 함수 반복) 만큼은 해치지 말고, 최대한 구현 시도를 하자
    ```
    - 만약 시간이 충분 했음에도 의견 조율에 실패했을 경우에 가져야할 태도
    ```
    1. 누구의 방식이 옳거나 그른 것은 없다
    2. 나의 방식을 마냥 고집하지 말고 각자의 의견을 잘 취합하여 프로젝트가 더 나은 방향으로 가려는 방식을 채택해야 한다
    3. 상대방의 방식을 존중하며 경청한다
    4. 나의 방식을 상대방에게 잘 전달하여 설득할 수 있는 능력을 길러야 한다
    ```
- 4일차 (목)
  - 구현 방식에 대해 토론했다 -> 합의
  - JS 작업
- 5일차 (금)
  - ...ing

## 미구현 사항

- 코드 리팩토링
- 노드 탐색 API 구현
- 정렬 탐색 알고리즘 구현

Commit:
docs : 리드미 파일 업데이트

리드미 파일 업데이트

bakhacode committed:
feat: 로딩시 변경이 없는 메인페이지 html,css 구현

간단하게 html과 css로 기본 틀을 먼저 잡았고, 버튼에 이벤트리스너만 달아두었다.

bakhacode committed:
docs: 리드미 수정

간단하게 순서를 나눠서 계획을 수립했다.

Commit:
docs : 리드미 파일 수정 및 업데이트

리드미 파일에 내용을 보충하고 기틀을 잡는다

Commit:
feat : 불필요한 html 요소 제거 및 css 속성 추가

불필요한 테스트용 html 요소를 제거하고 마을 지도의 border 색상을 변경한다

Commit:
feat: 마을을 생성하는 함수 구현

부모의 크기에 따라 위치와 크기가 제한되는 마을을 생성하는 함수를 구현한다

bakhacode committed:
feat: 마을지도 안에 마을을 여러개 생성하는 함수 구현

임의의 반복 횟수를 정해서 일단 마을이 생성이 잘 되는지 확인을 위한 함수를 구현했다.

namse commented on main.js:
```diff
@@ -0,0 +1,169 @@
+}
+
+function repeatMakeTown(outer) {
+    for (let i = 0; i < 3; i++) {
+        // if (outer.childNodes.length > 0) {
+        if (1) {
```
const exists = outer.childNodes.length > 0;
console.log('exists', exists);
if (exists) {

}
else {

}

Commit:
feat: html 태그명 변경 및 클래스명 변경

html의 태그를 용도에 맞게 변경하고 클래스명도 의도가 드러나게 변경한다

Commit:
style: css 작업 내용 수정

html요소의 패딩 및 마진 속성을 변경하여 가독성 좋게 만들고 버튼에 호버 기능을 추가하고 마을의 이름을 마을 왼편에 나타나게 한다

Commit:
feat: Town 클래스를 만들고 마을이 중첩되어 나타나도록 하는 함수 구현

Town 클래스를 만들어 마을의 이름, 크기, 위치 정보가 담긴 instance를 생성하게 하고 그 마을을 중첩되어 생성하게 하는 함수를 구현한다

Kakamotobi commented on .prettierrc:
```diff
@@ -0,0 +1,11 @@
+    {
+      "files": "*.js",
+      "options": {
+        "tabWidth": 4
```
Khundi - 특이한 설정이네요🤮

bakhacode replied:
JS에서는 4칸탭은 되야 보기에 편하더라구요... 나이가 들어서... 😢

Kakamotobi commented on index.html:
```diff
@@ -0,0 +1,26 @@
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <link rel="stylesheet" href="./src/style/main.css" />
+    <script type="module" src="./src/js/index.js"></script>
```
Khundi - `<head>`에 스크립트를 넣으면 오류 안나나요?

bakhacode replied:
defer랑 type="module"은 문제없는 걸로 알고있습니다!

aaaz425 replied:
type을 defer 또는 module로 설정하면 html 파싱 후에  script 파일을 실행해서 오류가 안나는 걸로 알고 있습니다!

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+class Town {
+    static townCount = 0;
+    constructor(outer) {
```
`outer`의 의미를 잘 모르겠어서 "parent"등 더 적절한 인자명을 쓰시면 좋을것 같아요.

bakhacode replied:
outer는 parent로 수정해야겠군요!

aaaz425 replied:
outer -> parent, inner -> children 으로 변경하면 더 명확하겠네요 감사합니다🫡

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+class Town {
+    static townCount = 0;
```
요구사항에는 없지만 `townCount`를 `Town` 클래스의 static property로 관리를 하면 (리프레시 안하고) 새로운 마을지도를 렌더링할 때 "A"부터 시작되지 않는 문제가 발생할 수 있을거 같네요.

bakhacode replied:
생각치못한 부분이였는데 감사합니다! 😲

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+        this.width = `${this.randomLength(outer.offsetWidth)}px`;
+        this.height = `${this.randomLength(outer.offsetHeight)}px`;
+        this.left = `${this.randomPoint(outer.offsetWidth)}px`;
+        this.top = `${this.randomPoint(outer.offsetHeight)}px`;
```
데이터를 문자열이 아닌 숫자값 그대로 저장하시는게 나중에 다른 곳에서 쓰기 더 적합할거 같아요.

bakhacode replied:
말씀대로 스타일 수정하는 부분에서 px이 들어가는게 맞는 것 같네용 😲

aaaz425 replied:
문자열로 저장하여 외부 함수에서 값을 사용할 때 parseInt를 사용하였는데, 숫자값으로 저장하면 의도도 명확해지고 활용하기에도 좋겠네요 !

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+        
+    }
+
+    randomPoint(outerLength) {
+        const centerPoint = outerLength / 2;
+        return Math.floor(Math.random() * centerPoint);
+    }
+
+    randomLength(outerLength) {
+        const minLength = 100;
+        const maxLength = outerLength / 2 - minLength;
+        return Math.floor(Math.random() * maxLength) + minLength;
+    }
```
Jello - 함수명, 인자명이 명확하지 않는거 같아요.🤔

bakhacode replied:
확실히 지금보니 구분해야할 필요성이 느껴지네요 좀 더 역할에 맡는 이름으로 수정해볼게요!

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+        return Math.floor(Math.random() * maxLength) + minLength;
+    }
+
+    appendInnerTown(outer) {
+        this.inner.textContent = this.name;
+        this.inner.style.left = this.left;
+        this.inner.style.top = this.top;
+        this.inner.style.width = this.width;
+        this.inner.style.height = this.height;
+        outer.appendChild(this.inner);
+    }
```
Jello - 함수명과 함수의 역할이 매칭이 안되는 것 같습니다 😵‍💫

bakhacode replied:
어펜드하는 동작과 스타일 수정하는 동작을 구분해서 분리해보겠습니다!

aaaz425 replied:
setChildStyle 으로 함수명을 정한다면 역할에도 맞고 의도도 명확해 질 것 같네요 감사합니다 🤗

Kakamotobi commented on src/js/index.js:
```diff
@@ -0,0 +1,76 @@
+            currentDepth++;
+        }
+    }
+}
```
공동 
- 불필요한 인자 및 인자 재할당으로 인해 혼란이 생기는것 같습니다.
- 함수명은 `createNestedTowns`가 더 적합할 것 같아요.

bakhacode replied:
함수에서 outer 매개변수 만들어놓고 인수로 전달안하면 선언만 한 상태로 유지된다는 거슬 배웠습니다! 그래서 이후에 outer 변수에 전달한 인수값이 아닌 함수내에서 새로운 값을 재할당해서 문제가 된다는 것을 인지완료~!

Kakamotobi commented on src/style/main.css:
```diff
@@ -0,0 +1,56 @@
+  display: flex;
+}
+
+h1 {
```
Khundi - 클래스명을 사용해주시면 좋을 것 같습니다 🖕
Geppetto - 앞에 id를 하나 적어주시면 좋을 것 같아요 🖕

bakhacode replied:
엄 격 하 게 클래스 사용할게요 ㅠ 😢

Kakamotobi commented on src/style/main.css:
```diff
@@ -0,0 +1,56 @@
+  font-size: 25px;
+}
+
+main div {
```
Jello - 셀렉터를 더 적절하게 사용하시면 더 좋을 것 같습니다.👍

bakhacode replied:
더 적절한 셀렉터 사용법을 찾아보겠습니다! 🤤

Commit:
feat: 길어진 함수를 분리

마을을 중첩하여 생성하는 함수를 분리하여 함수의 길이를 줄인다

bakhacode committed:
refactor: 변수명과 함수명 변경 및 함수 분리

모호한 의미의 변수명들을 변경해줬고, 동작에 맞게 함수명을 변경했다.
두 가지 이상의 동작을 하는 함수를 분리했다.

Commit:
feat: 마을이 겹치지 않게 생성되도록 하는 기능 사전 작업

마을이 겹치지 않게 생성되도록 하는 기능의 사전 작업을 한다

Commit:
docs: 리드미 업데이트

리드미에 필요 없는 내용을 정리하고 페어 프로그래밍 요약을 정리한다

bakhacode committed:
역할 전환

Commit:
style: 버튼을 클릭 했을 때 마을의 정보를 나타내는 p태그 수정 및 우체통 css 수정

html 요소 중 p태그의 텍스트를 비우고, 우체통을 가진 마을의 background 이미지를 설정한다

Commit:
feat: 우체통 이미지 파일 추가

우체통 이미지 파일을 추가한다

Commit:
feat: 버튼 클릭시 이벤트 추가

버튼을 클릭 했을 때 우체통을 가진 마을의 정보를 나타내고 border 색상을 변경하는 이벤트를 추가한다

Commit:
docs: 리드미 업데이트

완료된 체크리스트를 체크하고 필요없는 정보를 삭제하고 생각해볼거리를 추가한다

junzero741 commented on src/js/index.js:
```diff
@@ -0,0 +1,164 @@
+    }
+}
+
+function fn() {
```
함수 이름만 보고서는 어떤 일을 하는 함수인지 예측하기 어려울 수 있을 것 같습니다.

aaaz425 replied:
시간이 부족하여 아직 미완성된 함수 입니다!
여러 기능을 동시에 하고 있는 52번 줄의 createNestedTown 함수를 분리하여
서로 겹치지 않을 경우에만 새로운 중첩된 마을을 생성하는 기능을 담은 함수로 구현할 계획이었습니다
함수 이름은 아직 생각해보지 않았지만 addIsolatedTown 과 같이 함수명만 보고도 어떤 일을 하는 함수인지
예측할 수 있도록 더 고민해보도록 하겠습니다 !

junzero741 replied:
어떤 동작을 수행하는 함수를 A라고 했을 때, 
A의 본문 구현 -> A의 네이밍 
의 순서대로 개발을 진행하신게 아닌가 하는 생각이 드는데요.
이렇게 하면 자칫 개발 도중 내가 어떤 기능을 개발하고 있었는지 잊거나, 
A에서 할 일이 아닌데 A의 코드에 불필요한 코드가 들어갈 수 있습니다.
의식적으로 아래의 순서대로 코딩을 해보시는 걸 추천드려요.

A의 역할 정의 -> 역할을 잘 나타내는 A의 네이밍 정의 -> A의 본문 구현

junzero741 commented on src/js/index.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+function fn() {
+    createNestedTown();
+    createNestedTown();
+    createNestedTown();
+    createNestedTown();
+    createNestedTown();
```
이 함수가 딱 다섯번 실행되어야 하는 이유가 궁금합니다!

aaaz425 replied:
아직 미완성된 함수지만, 5번 실행 되는 동안 겹치지 않는 중첩된 마을들만 생성되는 함수입니다
다른 사람들을 납득 시킬 수 있을만한 조건과 변수들을 추가하여 구현해보겠습니다!

junzero741 replied:
5번이라는 횟수가 임의로 정한 것이든, 프로그램 요구 사항에 있는 것이든
따로 라벨을 달아서 저장해두는게 동료가 코드를 봤을때 혼동이 적을 것 같습니다.

코드의 반복이 일어나고 있는건 차치하더라도요.

junzero741 commented on src/js/index.js:
```diff
@@ -0,0 +1,164 @@
+            town.classList.add('mail-box');
+            const townWidth = town.offsetWidth;
+            const townHeight = town.offsetHeight;
+            town.id = townWidth * townHeight;
```
town 의 width 와 height 로 id를 만들고 있는데요, 모든 town 의 id 가 유니크한 값일까요?

aaaz425 replied:
우선 id 값을 비교해 마을의 크기 순서대로 정렬 하기 위해 
width와 height를 곱한 넓이를 id로 활용하였습니다!

만약 랜덤하게 생성된 town들의 넓이 값이 겹친다면 id 값이 더 이상 유니크 하지 않다는 점을 미쳐 생각 하지 못했습니다
넓이 값을 저장할 때, id가 아닌 다른 속성을 사용 하는 등 다른 방법을 찾아봐야 할 것 같습니다
감사합니다😀

bakhacode replied:
goody 안녕하세요! 리뷰에 대한 답변은 토코가 다 해주셔서 인사말 드립니다! 먼저 리뷰 요청을 위해 제대로 갖춰진 PR을 제출하지 못해서 죄송하다는 말씀드리고싶습니다...ㅠㅠ 어떤 핑계 댈 것 없이 능력 부족으로 준비를 못 했습니다.. 다음 주부터는 코드 정리와 함께 가이드에 맞춰 PR 작성 후 마감 늦지 않게 리뷰 요청드리겠습니다! 구디의 피드백으로 쑥쑥 자라는 모습 보여드릴 것을 약속드리고 4주 동안 잘 부탁드립니다~! 🔥🫡🔥

junzero741 replied:
토코 말씀대로, 넓이를 구한 이유가 마을의 크기 순서대로 정렬하기 위함이라면 아이디가 유니크하지 않을 수 있다는 문제 외에도, 아이디가 순서 정렬에 쓰이고 있다는 역할 문제도 있는 걸로 보입니다.
느끼신대로, 넓이 값을 저장할 다른 방법을 찾아보시는게 역시 좋을 것 같습니다. ㅎㅎ

junzero741 replied:
토코, 박하 두 분 다 안녕하세요. ㅎㅎㅎ
완성이 중요하게 느껴지시겠지만, 아직은 배우는 과정이니 완성하지 못해도 괜찮다고 생각해요.
죄송해하실 일도 아니구요!! 😌
미션 완수보다 함수 하나, 변수명 하나 충분히 고민하고 여러 시도를 해보시기를 권해드립니다. ㅎㅎ 앞으로 저도 잘 부탁드립니다!! 😊

junzero741 reviewed (approved):
안녕하세요, 이번 리뷰를 맡은 goody 라고 합니다!
만나서 반갑습니다. ㅎㅎ

전반적으로 함수들의 책임이 잘 나뉘어져 있는 것 같아서 좋았습니다.
자세한 리뷰는 코드에 따로 코멘트 남겨 두었습니다!

junzero741 commented:
.
